ding/policy/fqf.py

```diff
@@ -0,0 +1,294 @@
+        Returns:
+            - info_dict (:obj:`Dict[str, Any]`): Including current lr and loss.
+        """
+        data = default_preprocess_learn(
+            data, use_priority=self._priority, ignore_done=self._cfg.learn.ignore_done, use_nstep=True
+        )
+        if self._cuda:
+            data = to_device(data, self._device)
+        # ====================
+        # Q-learning forward
+        # ====================
+        self._learn_model.train()
+        self._target_model.train()
+        # Current q value (main model)
+        ret = self._learn_model.forward(data['obs'])
+        logit = ret['logit']  # [batch, 64(action_dim)]
+        q_value = ret['q']  # [batch, num_quantiles, 64(action_dim)]
+        quantiles = ret['quantiles']  # [batch, num_quantiles+1]
+        quantiles_hats = ret['quantiles_hats']  # [batch, num_quantiles]
+        q_tau_i = ret['q_tau_i']  # [batch_size, num_quantiles-1, action_size(64)]
+        entropies = ret['entropies']  # [batch, 1]
+
+        # Target q value
+        with torch.no_grad():
+            # quantiles_hats = (quantiles[:, :-1] + quantiles[:, 1:]) / 2.     # [batch_size, num_quantiles]
+            target_q_value = self._target_model.forward(data['next_obs'])['q']
+            # Max q value action (main model)
+            target_q_action = self._learn_model.forward(data['next_obs'])['action']
+
+            q_s_a_hats = evaluate_quantile_at_action(q_value, data['action'])  # [batch_size, num_quantiles, 1]
+
+        data_n = fqf_nstep_td_data(
+            q_value, target_q_value, data['action'], target_q_action, data['reward'], data['done'], quantiles_hats,
+            data['weight']
+        )
+        value_gamma = data.get('value_gamma')
+
+        entropy_loss = -self._ent_coef * entropies.mean()
+        
+        fraction_loss = fqf_calculate_fraction_loss(q_tau_i.detach(), q_s_a_hats.detach(), quantiles, data['action']) + entropy_loss
+
+        quantile_loss, td_error_per_sample = fqf_nstep_td_error(
+            data_n, self._gamma, nstep=self._nstep, kappa=self._kappa, value_gamma=value_gamma
+        )
+
+        # ====================
+        # fraction_proposal network update
+        # ====================
+        self._fraction_loss_optimizer.zero_grad()
+        fraction_loss.backward(retain_graph=True)
+        if self._cfg.learn.multi_gpu:
+            self.sync_gradients(self._learn_model)
+        total_norm_quantiles_proposal = torch.norm(
+            torch.stack(
+                [torch.norm(p.grad.detach(), 2.0) for p in self._model.head.quantiles_proposal.parameters()]
+            ), 2.0
+        )
+        self._fraction_loss_optimizer.step()
+
+        # ====================
+        # Q-learning update
+        # ====================
+        self._quantile_loss_optimizer.zero_grad()
+        quantile_loss.backward()
+        if self._cfg.learn.multi_gpu:
+            self.sync_gradients(self._learn_model)
+        total_norm_Q = torch.norm(
```

Comment:
abstract a function to calculate norm, add all the procedure should be in `torch.no_grad` context for speeding up